proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+* Prepare the compiler internals for further work on dependent types
+
+Proposal Structure
+------------------
+
+We shall present this proposal in two parts:
+
+* In Part I we introduce the ``forall a ->`` quantifier in types of terms but
+  also require a syntactic marker at use sites. This is not as convenient to
+  use (i.e. users would have to write ``sizeOf (type Bool)`` instead of
+  ``sizeOf Bool``), but is much easier to specify and understand.
+
+* In Part II we specify when it is permissible to omit the ``type`` herald.
+  This greatly increases the convenience of using the proposed feature, but
+  also makes the specification more intricate.
+
+
+Part I: Proposed Change Specification
+-------------------------------------
+
+1. Add a new language extension, ``RequiredTypeArguments``. When
+   ``RequiredTypeArguments`` is in effect, lift the restriction that the
+   ``forall a ->`` quantifier cannot be used in types of terms.
+
+2. **Syntax**. When ``ExplicitNamespaces`` is in effect, extend the
+   grammar as follows::
+
+        exp ::=
+          | 'type' type
+          | ...
+
+        pat ::=
+          | 'type' type
+          | ...
+
+3. **Name resolution**. A type embedded into a term with the ``type`` marker
+   follows type-level name resolution rules (i.e. uses of punned identifiers
+   resolve to the type namespace).
+
+   The ``ScopedTypeVariables`` extension has no effect on variables introduced
+   by ``forall a ->``.
+
+4. **Type checking**. In type checking, we alternate between two
+   distinct modes: *checking* and *inference*. This idea, called bidirectional
+   type checking, is presented in more detail in
+   `"A quick look at impredicativity" <https://www.microsoft.com/en-us/research/uploads/prod/2020/01/quick-look-icfp20.pdf>`_.
+
+   * In inference mode, we never infer ``forall x -> t``.
+
+   * In checking mode, in a function application chain ``f e1 e2 e3``, we
+     follow the rules shown in Figure 4 of "A quick look at impredicativity",
+     extended as follows::
+
+        G |- sigma[a := rho]; pis  ~>  Theta; phis; rho_r
+        ------------------------------------------------------------------  ITVDQ
+        G |- (forall a -> sigma); (type rho), pis  ~>  Theta; phis; rho_r
+
+   * In checking mode, in a function binding ``f (type x) = ...`` or a lambda
+     ``\(type x) -> ...``, the ``x`` is a fresh skolem.
+
+5. **Validity**. Expressions and patterns of form ``type t`` but not covered by
+   the type checking rules above are illegal.
+
+   Specifically, any expression of form ``type t`` must be used as an argument
+   to a function, or else it is rejected with a type error::
+
+     x = f (type Int)   -- OK
+     x = type Int       -- invalid use of a type in a term
+
+   This is checked during type checking, so Template Haskell is unaffected, and
+   ``[e| type Int |]`` is allowed (but different from ``[t| Int |]``).
+
+   Furthermore, any pattern of form ``type t`` must be either a variable or a
```

Comment:
What about

```hs
data T a where
  MkT :: forall a -> a -> T a

f (MkT (type Int) x) = x + 5
```

?

Reply:
if we don't have a separate extension for invisible type patterns, then I think we get that for free. But note `type Int` has to be a pattern in its own right, `type boundVariable` is no good per discussion in #291.